src/intptrcast.rs

```diff
@@ -179,6 +195,9 @@ impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir,
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
     fn expose_ptr(&mut self, alloc_id: AllocId, tag: BorTag) -> InterpResult<'tcx> {
         let ecx = self.eval_context_mut();
+        // Make sure we aren't trying to expose a dead allocation; this is part of justifying the
+        // way we treat the exposed set in VisitProvenance.
+        assert!(!matches!(ecx.get_alloc_info(alloc_id).2, AllocKind::Dead));
```

Comment:
Exposing a dead allocation is totally legal. It just shouldn't add the allocation to `exposed`, since you can't int2ptr cast it anyway.